### DISCUSSION TITLE
WMS Style Problem

### DISCUSSION THREAD

**Discussion by marcoma9023** [pelicanmapping/osgearth#2330] [Support Q&A] [answered]
Greetings,

I have a WMS for which I need to change the style at runtime.

I would like to know, even in terms of cache, which is the most correct approach to avoid slow tile processing.

1) One layer for each style and associated with it a bin for the cache
2) A single layer, with only one bin
3) A single layer and a bin for each style

Currently I'm handling it like point 1), so, every time I change the style, I remove the layer from the Map and put the one with the new style on it.

However, this approach causes me problems of slowness when I remove and add, in fact it seems that the callback freezes, the earth disappears, and after about 10 seconds it resumes and manages the new tiles.

I also tried to flag setAsyncLoading(true), but it doesn't improve the situation.

Thanks for the support.

**Comment by gwaldron** [comment, marked as answer]
I'm not sure if this will work for WMS, but instead of removing/adding layers you can also try the invalidate API:
```c++
std::vector<const Layer*> layers;
layers.push_back(your_wms_layer);
GeoExtent extent;
mapNode->getTerrainEngine()->invalidateRegion(layers, extent);
```
This tells the terrain engine to re-load all the tiles from those layers in that extent.
